src/backend/access/aocs/aocssegfiles.c

```diff
@@ -947,11 +948,15 @@ UpdateAOCSFileSegInfo(AOCSInsertDesc idesc)
 									&null[Anum_pg_aocs_vpinfo - 1]);
 
 	Assert(!null[Anum_pg_aocs_vpinfo - 1]);
-	struct varlena *v = (struct varlena *) DatumGetPointer(d_tmp);
-	struct varlena *dv = pg_detoast_datum(v);
 
-	Assert(VARSIZE(dv) == aocs_vpinfo_size(nvp));
-	AOCSVPInfo *oldvpinfo = (AOCSVPInfo *) dv;
+	AOCSFileSegInfo *seginfo;
+	seginfo = (AOCSFileSegInfo *) palloc0(aocsfileseginfo_size(nvp));
+
+	deformAOCSVPInfo(prel,
+					 (struct varlena *) DatumGetPointer(d_tmp),
+					 seginfo);
+
+	AOCSVPInfo *oldvpinfo = (AOCSVPInfo *) &seginfo->vpinfo;
```

Comment:
does this change more related to the previous commit?
We also don't need to cast seginfo->vpinfo to AOCSVPInfo